BUG: view with fieldless dtype should raise if itemsize != 0

ahaldane opened the pull request:
Fixes #14344

Minor bugfix caused by treating the "dtype([])" structured type as if it was the "V0" flexible type.

ahaldane committed:
BUG: view with fieldless dtype should raise if itemsize != 0

Fixes #14344

eric-wieser commented:
I think this now warrants a few more tests, since this fixed a lot of bugs.
Here's another that I think is fixed
```
>>> np.dtype((np.dtype([]), 10))
dtype({'names':[], 'formats':[], 'offsets':[], 'itemsize':10})  # before
dtype(([], (10,)))  # after
```

eric-wieser commented:
Probably worth a `..versionchanged` in `doc/source/reference/c-api/array.rst` too, where this macro is documented.

eric-wieser commented:
Another fix, `np.fromiter((() for i in range(10)), [])` doesn't raise

ahaldane committed:
BUG: PyDataType_ISUNSIZED should be False for structured types

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1200,7 +1200,8 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
+            /* these dtypes had no fields */
+            return cmp_op == Py_EQ ? Py_True : Py_False;
```
This change isn't correct - it doesn't produce a correctly broadcast array. Let's leave this to a future PR - the rest looks great.

ahaldane replied:
Fixed by falling through to the string compare case which does the right thing here.

ahaldane commented:
Thanks for great reviews, as usual.

I updated `PyDataType_ISUNSIZED`, and went through all the places it was used to check if this introduced any bugs. It did, which I fixed, and I added appropriate tests.

I have some hesitation on doing it this way, so consider this a WIP. First, in some of the places calling `ISUNSIZED` the code really did want to check just `elsize == 0`, fields or no fields. Also, with the same logic it seems like `PyDataType_ISFLEXIBLE` should also be False for structured 0-size types. However, when I make that change I get a bunch of segfaults, especially in the pickling code. .

ahaldane committed:
BUG: Fix comparison of fieldless structured arrays

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1199,15 +1199,13 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
-         * compare as a string. Assumes self and
-         * other have same descr->type
+         * compare as a string. Assumes self and other have same descr->type.
+         * Note that fieldless structured datatypes (itemsize 0) fall through
```
What about fieldless types with padding?

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1200,15 +1200,18 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
+            /* these dtypes had no fields */
+            res = PyArray_NewLikeArray(self, NPY_ANYORDER,
```
This doesn't broadcast with `other`, unless the broadcasting has already happened here

ahaldane replied:
Ah, yeah. I think I've got it finally by using a multiiter.

ahaldane committed:
BUG: Fix fieldless comparison broadcasting

Fixes #13438

eric-wieser commented on numpy/core/src/multiarray/arrayobject.c:
```diff
@@ -1200,15 +1200,28 @@ _void_compare(PyArrayObject *self, PyArrayObject *other, int cmp_op)
             }
         }
         if (res == NULL && !PyErr_Occurred()) {
-            PyErr_SetString(PyExc_ValueError, "No fields found.");
```
Fixes gh-13438, I think

eric-wieser commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -1861,7 +1861,7 @@ array_reduce_ex(PyArrayObject *self, PyObject *args)
         (PyType_IsSubtype(((PyObject*)self)->ob_type, &PyArray_Type) &&
          ((PyObject*)self)->ob_type != &PyArray_Type) ||
-        PyDataType_ISUNSIZED(descr)) {
+        descr->elsize == 0) {
```
It's not clear to me why this condition is needed at all, although it admittedly looks harmless.

ahaldane replied:
not clear to me either. These lines previously discussed here: https://github.com/numpy/numpy/pull/12748/files#r249258808 and introduced in #12011, but not discussed there.

eric-wieser commented on doc/source/reference/c-api/array.rst:
```diff
@@ -997,6 +997,9 @@ argument must be a :c:type:`PyObject *<PyObject>` that can be directly interpret
     be sized, hence the array form of this macro not existing.
 
+    ..versionchanged:: 1.18.0
+    For structured datatypes with no fields this function now return False.
```
Does this line need to be indented?

typo: `returns`

ahaldane replied:
Fixed typo. I indented it the same way as the `..versionadded` in this file so it should be ok.

eric-wieser reviewed (approved):
Looks pretty good - one unimportant question, and a doc typo. Getting the comparison fix in here should have enabled me to push some of my remaining fixes, which were untestable without it - thanks!

eric-wieser commented:
I think an `upcoming_changes/14393.c_api.rst` note might be warranted about the macro behavior

charris commented:
@ahaldane Ping.

ahaldane commented:
About to push as you wrote that!

mattip commented:
Thanks @ahaldane and reviewers
